File: apps/web/app/application/use-cases/register-client.ts
import type { Client } from '~/domain/entities/client'
import type { ClientRepository, RegisterClientInput } from '~/application/ports/client-repository'

export class RegisterClient {
  constructor(private readonly clients: ClientRepository) {}

  execute(input: RegisterClientInput): Promise<Client> {
    return this.clients.register(input)
  }
}

File: apps/server/src/server/application/use_cases/register_client.py
import uuid
from dataclasses import dataclass
from datetime import UTC, datetime

from server.domain.entities import Client
from server.domain.ports import ClientRepository


@dataclass(frozen=True, slots=True)
class RegisterClientInput:
    name: str
    tax_id: str | None
    email: str | None


class RegisterClient:
    def __init__(self, clients: ClientRepository) -> None:
        self._clients = clients

    def execute(self, data: RegisterClientInput) -> Client:
        client = Client(
            id=str(uuid.uuid4()),
            name=data.name,
            tax_id=data.tax_id,
            email=data.email,
            created_at=datetime.now(UTC),
        )
        self._clients.save(client)
        return client
